# Financial Fraud Detection & Risk Intelligence System

## 01 — Data Audit

### Objective

Understand the structure, quality, completeness, and reliability
of the financial transaction dataset before exploratory data
analysis and machine learning.

### Dataset

Financial Fraud Detection Dataset

### Target Variable

Fraudulent

- 0 = Non-fraud
- 1 = Fraud

In [4]:
import pandas as pd
import numpy as np

In [5]:
# Load raw dataset

df = pd.read_csv(
    "../data/raw/financial_fraud_detection_dataset.csv"
)

print("Dataset loaded successfully")

Dataset loaded successfully


In [6]:
# Dataset shape

print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 5000
Number of columns: 14


In [7]:
# First five rows
df.head()

# Last five rows
df.tail()

# Random sample
df.sample(5, random_state=42)

,Transaction_ID,Customer_ID,Transaction_Date,Transaction_Amount,Merchant_Category,Payment_Method,Device_Type,Location,Is_International,Previous_Transactions,Average_Spend,Account_Age_Days,Suspicious_Keyword,Fraudulent
1501,T101501,CUST7613,06-02-2023 03:10,52.19,Travel,UPI,Desktop,Bengaluru,1,154,157.70,1237,No,1
2586,T102586,CUST4507,16-10-2023 16:45,5.96,Grocery,UPI,Mobile,Hyderabad,0,135,431.73,232,No,0
2653,T102653,CUST8399,23-04-2023 18:19,4.30,Electronics,UPI,Desktop,Chennai,0,136,265.98,869,No,0
1055,T101055,CUST1793,29-04-2023 11:56,34.55,Grocery,Credit Card,Desktop,Delhi,0,6,304.56,791,No,0
705,T100705,CUST7605,26-05-2023 15:10,368.10,Health,PayPal,POS,Pune,1,182,328.05,1123,No,0


In [9]:
# Dataset information
df.info()

# Column names
print(df.columns.tolist())

<class 'pandas.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Transaction_ID         5000 non-null   str    
 1   Customer_ID            5000 non-null   str    
 2   Transaction_Date       5000 non-null   str    
 3   Transaction_Amount     5000 non-null   float64
 4   Merchant_Category      5000 non-null   str    
 5   Payment_Method         5000 non-null   str    
 6   Device_Type            5000 non-null   str    
 7   Location               5000 non-null   str    
 8   Is_International       5000 non-null   int64  
 9   Previous_Transactions  5000 non-null   int64  
 10  Average_Spend          5000 non-null   float64
 11  Account_Age_Days       5000 non-null   int64  
 12  Suspicious_Keyword     5000 non-null   str    
 13  Fraudulent             5000 non-null   int64  
dtypes: float64(2), int64(4), str(8)
memory usage: 844.6 KB
['Transactio

In [10]:
# Missing-value summary

missing_summary = pd.DataFrame({
    "Missing_Count": df.isnull().sum(),
    "Missing_Percentage": (
        df.isnull().sum() / len(df) * 100
    ).round(2)
})

print(missing_summary)

                       Missing_Count  Missing_Percentage
Transaction_ID                     0                 0.0
Customer_ID                        0                 0.0
Transaction_Date                   0                 0.0
Transaction_Amount                 0                 0.0
Merchant_Category                  0                 0.0
Payment_Method                     0                 0.0
Device_Type                        0                 0.0
Location                           0                 0.0
Is_International                   0                 0.0
Previous_Transactions              0                 0.0
Average_Spend                      0                 0.0
Account_Age_Days                   0                 0.0
Suspicious_Keyword                 0                 0.0
Fraudulent                         0                 0.0


In [12]:
# Complete duplicate rows
print("Duplicate rows:", df.duplicated().sum())

# Duplicate transaction IDs
print(
    "Duplicate Transaction IDs:",
    df["Transaction_ID"].duplicated().sum()
)

Duplicate rows: 0
Duplicate Transaction IDs: 0


In [13]:
# Unique values per column

unique_summary = pd.DataFrame({
    "Column": df.columns,
    "Unique_Values": df.nunique().values
})

print(unique_summary)

                   Column  Unique_Values
0          Transaction_ID           5000
1             Customer_ID           3847
2        Transaction_Date           4980
3      Transaction_Amount           4302
4       Merchant_Category              8
5          Payment_Method              5
6             Device_Type              3
7                Location              7
8        Is_International              2
9   Previous_Transactions            199
10          Average_Spend           4753
11       Account_Age_Days           1827
12     Suspicious_Keyword              2
13             Fraudulent              2


In [14]:
# Data types
print(df.dtypes)

Transaction_ID               str
Customer_ID                  str
Transaction_Date             str
Transaction_Amount       float64
Merchant_Category            str
Payment_Method               str
Device_Type                  str
Location                     str
Is_International           int64
Previous_Transactions      int64
Average_Spend            float64
Account_Age_Days           int64
Suspicious_Keyword           str
Fraudulent                 int64
dtype: object


In [15]:
# Target distribution

target_summary = (
    df["Fraudulent"]
    .value_counts()
    .sort_index()
    .reset_index()
)

target_summary.columns = ["Fraudulent", "Count"]

target_summary["Percentage"] = (
    target_summary["Count"] / len(df) * 100
).round(2)

print(target_summary)

   Fraudulent  Count  Percentage
0           0   4518       90.36
1           1    482        9.64


In [16]:
# Identifier uniqueness

identifier_columns = [
    "Transaction_ID",
    "Customer_ID"
]

for column in identifier_columns:
    print(
        column,
        "Unique values:",
        df[column].nunique(),
        "Total rows:",
        len(df)
    )

Transaction_ID Unique values: 5000 Total rows: 5000
Customer_ID Unique values: 3847 Total rows: 5000


In [17]:
# Numerical feature summary

print(df.describe())

       Transaction_Amount  Is_International  Previous_Transactions  \
count         5000.000000       5000.000000            5000.000000   
mean            79.180664          0.088200              99.414000   
std             78.964045          0.283614              57.022968   
min              0.000000          0.000000               1.000000   
25%             22.365000          0.000000              50.000000   
50%             55.455000          0.000000             100.000000   
75%            110.297500          0.000000             148.000000   
max            653.800000          1.000000             199.000000   

       Average_Spend  Account_Age_Days   Fraudulent  
count    5000.000000       5000.000000  5000.000000  
mean      258.377132       1011.158200     0.096400  
std       140.230639        568.633834     0.295169  
min        10.010000         30.000000     0.000000  
25%       138.772500        530.000000     0.000000  
50%       261.810000       1012.500000     0.

In [18]:
# Categorical columns

categorical_columns = [
    "Merchant_Category",
    "Payment_Method",
    "Device_Type",
    "Location",
    "Is_International",
    "Suspicious_Keyword"
]

for column in categorical_columns:
    print(f"\n{column}")
    print(df[column].value_counts(dropna=False))


Merchant_Category
Merchant_Category
Food             684
Electronics      650
Entertainment    630
Grocery          629
Fashion          608
Health           606
Travel           599
Utilities        594
Name: count, dtype: int64

Payment_Method
Payment_Method
NetBanking     1037
PayPal         1022
Debit Card     1008
Credit Card     972
UPI             961
Name: count, dtype: int64

Device_Type
Device_Type
Desktop    1679
POS        1673
Mobile     1648
Name: count, dtype: int64

Location
Location
Delhi        758
Mumbai       741
Hyderabad    725
Pune         722
Kolkata      713
Bengaluru    686
Chennai      655
Name: count, dtype: int64

Is_International
Is_International
0    4559
1     441
Name: count, dtype: int64

Suspicious_Keyword
Suspicious_Keyword
No     4740
Yes     260
Name: count, dtype: int64


In [19]:
# Convert date for validation

df["Transaction_Date"] = pd.to_datetime(
    df["Transaction_Date"],
    format="%d-%m-%Y %H:%M",
    errors="coerce"
)

print("Invalid dates:", df["Transaction_Date"].isna().sum())

print("Minimum date:", df["Transaction_Date"].min())
print("Maximum date:", df["Transaction_Date"].max())

Invalid dates: 0
Minimum date: 2023-01-01 02:16:00
Maximum date: 2024-02-21 15:34:00


## Data Audit — Key Findings

1. The dataset contains 5,000 transaction records.
2. The target variable is Fraudulent.
3. The dataset contains 482 fraud records and 4,518 non-fraud records.
4. No missing values were detected during the audit.
5. No complete duplicate rows were detected.
6. Transaction_ID is a unique identifier.
7. Customer_ID is repeated across some transactions.
8. Transaction_Date requires date parsing using the expected format.
9. Suspicious_Keyword requires a separate leakage review.
10. The dataset should be documented as synthetic.